resin/logs.py

```diff
@@ -18,104 +21,50 @@ class Logs(object):
+    def subscribe(self, uuid, callback):
         """
         This function allows subscribing to device logs.
-        There are fields (`m`, `t`, `s`, `c`) in the output which can be unclear. They stand for:
-            m - The log message itself.
-            t - The log timestamp.
-            s - Is this a system message?
-            c - The id of the service which produced this log (or null if the device does not support multiple containers).
 
         Args:
             uuid (str): device uuid.
             callback (function): this callback is called on receiving a message from the channel.
-            error (function): this callback is called on an error event.
-            For more details about callbacks in pubnub subscribe, visit here: https://www.pubnub.com/docs/python/api-reference#subscribe
-
-        Examples:
-            # Define callback and error.
-            >>> def callback(message, channel):
-            ...     print(message)
-            >>> def error(message):
-            ...     print('Error:'+ str(message))
-            >>> Logs.subscribe(uuid=uuid, callback=callback, error=error)
 
         """
 
-        channel = self.get_channel(uuid)
-        self.pubnub.subscribe(channels=channel, callback=callback, error=error)
-
-    @_init_pubnub
-    def history(self, uuid, callback, error):
+        raw_query = 'stream=1'
+        response = self.base_request.request(
+            '/device/v2/{uuid}/logs'.format(uuid=uuid), 'GET', raw_query=raw_query, stream=True,
+            endpoint=self.settings.get('api_endpoint')
+        )
+
+        if (response.status_code == 200):
+            for line in response.iter_lines(chunk_size=None):
+                if line:
+                    entry = json.loads(line)
+                    callback(entry)
+        else:
+            raise exceptions.RequestError(response.content)
+
+    def history(self, uuid, count=None):
         """
-        This function allows fetching historical device logs.
-
+        Get device logs history.
         Args:
             uuid (str): device uuid.
-            callback (function): this callback is called on receiving a message from the channel.
-            error (function): this callback is called on an error event.
-            For more details about callbacks in pubnub subscribe, visit here: https://www.pubnub.com/docs/python/api-reference#history
-
-        Examples:
-            # Define callback and error.
-            >>> def callback(message):
-            ...     print(message)
-            >>> def error(message):
-            ...     print('Error:'+ str(message))
-            Logs.history(uuid=uuid, callback=callback, error=error)
+            count (int): this callback is called on receiving a message from the channel.
 
         """
 
-        channel = self.get_channel(uuid)
-        self.pubnub.history(channel=channel, callback=callback, error=error)
-
-    def unsubscribe(self, uuid):
```

Comment:
This has been removed, and now there's no way for users to unsubscribe at all, right? That seems pretty inconvenient both for end users and us, if it means we end up keeping lots of unwanted connections open.

Looks like it should be fairly easy to implement something to do this with just `response.close()`. An API design that allows that is a trickier question though... In the JS SDK, we return an object from subscribe, with `on('line', (line) => ...)` methods to listen to events, plus an `unsubscribe()` method to kill the subscription & connection entirely.

I'm not really up to date on idiomatic Python, but I imagine something similar might fit well here.

Reply:
Will add this asap.